### ISSUE TITLE
DOC: complex conjugation in BLAS functions

### ISSUE THREAD

**Issue by rileyjmurray** [scipy/scipy#15145]
### Is your feature request related to a problem? Please describe.

I have a very large complex array ``A`` and I want to apply its adjoint (conjugate transpose) to a vector ``y``. If I use ``A.conj().T @ y`` then ``A.conj()`` is allocated explicitly.  That consumes too much memory. It's possible to bypass allocation by calling ``scipy.linalg.blas.zgemv(1.0, A, y, trans=2)``.

The trouble is: the meaning of the ``trans`` parameter isn't specified in SciPy's [zgemv](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.blas.zgemv.html#scipy-linalg-blas-zgemv) documentation. I only learned of this solution by guessing the behavior of ``trans`` and trying it out.

### Describe the solution you'd like.

SciPy's documentation for complex BLAS functions should indicate how to perform the conjugate-transpose operation. I.e., explain the meanings of trans = 0, 1, or 2 in documentation for the relevant functions.

I think the change is reasonable because people who are familiar with common C or FORTRAN BLAS interfaces expect trans to be a character (N, T, or H).

### Describe alternatives you've considered.

N/A

### Additional context (e.g. screenshots, GIFs)

Code snippet showing undocumented functionality:
```
import numpy as np
import scipy.linalg as la
import scipy.linalg.blas as blas

A = np.random.randn(3,2) + 1j*np.random.randn(3,2)
y = np.random.randn(3) + 1j

inefficient = A.T.conj() @ y
efficient = blas.zgemv(1.0, A, y, trans=2)
print(la.norm(inefficient - efficient))
```

**Comment by ilayn** [comment]
Hi @rileyjmurray The BLAS and LAPACK function in SciPy are python wrappers around the backend of your choice, MKL, OpenBLAS and so on. And historically the wrappers are used with integers instead of characters. So whatever option is on netlib side they are all counted as integers. It is unfortunately a bit late to do anything about it without breaking lots of code out there.  The surgery on the documentation can maybe be done with some effort but I wouldn't know where to start being perfectly honest.

**Comment by rileyjmurray** [comment]
Thanks for the reply @ilayn! If I understand what you're saying, then my initial post saying "``trans=2`` is conjugate transpose" could potentially be incorrect for people who linked to different versions of BLAS?

**Comment by ilayn** [comment]
Ah no, fortunately we are safe in that regard that BLAS functions are almost written on stone after decades of usage. What I meant is if a keyword has options `A, B, C` for some reason happened before I was around, they are counted as 0, 1, 2 and this happens quite a number of routines. And in some of them string chars are used literraly. So you are right that we are missing consistency across the whole routine catalog.

**Comment by rileyjmurray** [comment]
Great. So, what are some possibilities for moving forward here? Is there a brute-force option where someone (e.g., me) could expands a couple dozen documentation files with a suitable phrase for complex conjugation? I'd suggest something like "Here, setting trans = 0, 1, 2 correspond to trans = 'N', 'T', 'H' in the Netlib LAPACK implementation." You seem to also suggest that modifying f2py might work. Who could we talk to to figure out if that's actually viable?

**Comment by melissawm** [comment]
I am taking a look into this - the difficulty is that these are autogenerated modules, and the documentation is also autogenerated from the wrappers, without any extra info about the parameters that f2py can parse. I don't see any immediate solutions other than being able to use a comment as extra info for the docstring.

**Comment by mdhaber** [comment]
I think we could have a standard note, something like:
> Complete documentation for all parameters listed here is available at <link to Netlib, possibly for the particular function>; other parameters are computed internally and are not required when calling the function. Note that historically, BLAS/LAPACK wrappers accepted integers where characters are now expected. Should the SciPy wrapper accepts an integer and the Netlib documentation document only characters, please pass an integer corresponding with the desired option: 0 for the first option listed, 1 for the second, etc.

(Not sure if that rule is even generally correct; another option is to list common conversions.)

I see three obvious times/places where this sort of note could be injected:
1. By `f2py` when the wrapper is generated. Hardest to figure out, but fastest and always available.
2. With a loop in `lapack` at import time. Slows import, but available when imported directly (e.g. `help(scipy.linalg.blas.zgemv)`) AND via `get_X_funcs` (e.g. `help(scipy.linalg.get_lapack_funcs('zgemv', np.complex128)`). I'm guessing it would render online, too.
3. By `get_lapack_funcs`/`get_blas_funcs`. Does not slow import, just retrieval using `get_X_funcs`. Not available when imported directly.

@melissawm Do you see a place we could add such a standard note when generating the wrappers?

If not, suggestions about 2, 3, or something more exotic? I could implement either of those two if we think a standard note would be better than nothing.

**Comment by ev-br** [comment]
Addressing this issue could be a tad easier now that the BLAS wrapping is no longer using f2py.

**Comment by ilayn** [comment]
Yes now all docs are customizable albeit a bit awkward C++ silliness involved to keep the binary size small.
